packages/http/src/body.ts

```diff
@@ -0,0 +1,144 @@
+  requestOrResponseType: Type,
+  metadata: Set<ModelProperty>,
+  rootPropertyMap: Map<ModelProperty, ModelProperty>,
+  visibility: Visibility
+): [ResolvedBody | undefined, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  // non-model or intrinsic/array model -> response body is response type
+  if (requestOrResponseType.kind !== "Model" || isArrayModelType(program, requestOrResponseType)) {
+    return diagnostics.wrap({ type: requestOrResponseType, isExplicit: false });
+  }
+
+  const duplicateTracker = new DuplicateTracker<string, Type>();
+
+  // look for explicit body
+  let resolvedBody: ResolvedBody | undefined;
+  for (const property of metadata) {
+    const isBodyVal = isBody(program, property);
+    const isBodyRootVal = isBodyRoot(program, property);
+    if (isBodyVal || isBodyRootVal) {
+      duplicateTracker.track("body", property);
+      if (resolvedBody === undefined) {
+        resolvedBody = { type: property.type, isExplicit: isBodyVal, property };
+      }
+    }
+  }
+  for (const [_, items] of duplicateTracker.entries()) {
+    for (const prop of items) {
+      diagnostics.add(
+        createDiagnostic({
+          code: "duplicate-body",
+          target: prop,
+        })
+      );
+    }
+  }
+  if (resolvedBody === undefined) {
+    // Special case if the model as a parent model then we'll return an empty object as this is assumed to be a nominal type.
+    // Special Case if the model has an indexer then it means it can return props so cannot be void.
+    if (requestOrResponseType.baseModel || requestOrResponseType.indexer) {
+      return diagnostics.wrap({ type: requestOrResponseType, isExplicit: false });
+    }
+    // Special case for legacy purposes if the return type is an empty model with only @discriminator("xyz")
+    // Then we still want to return that object as it technically always has a body with that implicit property.
+    if (
+      requestOrResponseType.derivedModels.length > 0 &&
+      getDiscriminator(program, requestOrResponseType)
+    ) {
+      return diagnostics.wrap({ type: requestOrResponseType, isExplicit: false });
+    }
+  }
+
+  const bodyRoot = resolvedBody?.property ? rootPropertyMap.get(resolvedBody.property) : undefined;
+
+  const unannotatedProperties = filterModelProperties(
+    program,
+    requestOrResponseType,
+    (p) => !metadata.has(p) && p !== bodyRoot && isVisible(program, p, visibility)
+  );
+
+  if (unannotatedProperties.properties.size > 0) {
+    if (resolvedBody === undefined) {
+      return diagnostics.wrap({ type: unannotatedProperties, isExplicit: false });
+    } else {
+      diagnostics.add(
+        createDiagnostic({
+          code: "duplicate-body",
+          messageId: "bodyAndUnannotated",
+          target: requestOrResponseType,
+        })
+      );
+    }
+  }
+
+  return diagnostics.wrap(resolvedBody);
+}
+
+/** Validate a property marked with `@body` */
+export function validateBodyProperty(
```

Comment:
decide if we do want to emit a warning if the metadata wasn't applicable to start with `@path` or `@query` in response for example. Currently we do but if this already was ignored with implicit body then not sure we should complain as there is no effect